### PR TITLE
Enhancement - Fix preprocessing crash when num_workers=0 in Megatron GPT3 dataset generation

### PR DESCRIPTION
**Pull request by polarG**
**Description**
preprocess_data.py uses multiprocessing.Pool(workers), which requires workers >= 1. When num_workers is set to 0 (valid for DataLoader, where it means "load in main process"), the preprocessing step crashes.

This change clamps the worker count to max(1, self._args.num_workers) before passing it to preprocess_data.py, while leaving the original num_workers value unchanged for other uses like DataLoader.

### CODE REVIEW

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 1]
```diff
@@ -651,13 +651,16 @@ def _generate_dataset(self):
...
                         'python3 '
                         f'{os.path.join(self._args.code_base, "tools/preprocess_data.py")} '
                         f'--input {self._raw_data_path} '
                         f'--tokenizer-type {self._args.tokenizer_type} '
                         f'--output-prefix {os.path.join(self._args.data_home, "dataset")} '
-                        f'--workers {str(self._args.num_workers)} '
+                        # num_workers=0 is valid for DataLoader (main process loads data),
+                        # but preprocess_data.py requires workers>=1 for multiprocessing.Pool.
+                        f'--workers {max(1, self._args.num_workers)} '
                         f'--vocab-file {self._vocab_path} '
```
[the remark this answers is not in the document]
Comment (polarG) [reply superbench/benchmarks/model_benchmarks/megatron_gpt3.py:707]: Addressed in c6d205e4: derive `--output-prefix` from `data_prefix` by stripping a trailing `_text_document` suffix when present (e.g., default `dataset_text_document` -> `<data_home>/dataset`; `custom_text_document` -> `<data_home>/custom`; a non-suffixed `mydata` is used as-is). This keeps generation and the subsequent `{data_prefix}.bin/.idx` existence checks consistent for any custom value.

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 2]
```diff
@@ -651,13 +651,16 @@ def _generate_dataset(self):
...
                     command = (
                         'python3 '
                         f'{os.path.join(self._args.code_base, "tools/preprocess_data.py")} '
                         f'--input {self._raw_data_path} '
                         f'--tokenizer-type {self._args.tokenizer_type} '
                         f'--output-prefix {os.path.join(self._args.data_home, "dataset")} '
-                        f'--workers {str(self._args.num_workers)} '
+                        # num_workers=0 is valid for DataLoader (main process loads data),
+                        # but preprocess_data.py requires workers>=1 for multiprocessing.Pool.
+                        f'--workers {max(1, self._args.num_workers)} '
```
[the remark this answers is not in the document]
Comment (polarG) [reply superbench/benchmarks/model_benchmarks/megatron_gpt3.py]: Addressed in c6d205e4 (substring match tightened in ee2839e9): added `test_megatron_gpt_dataset_generate_command` covering (1) `num_workers=0` with default `data_prefix` -> `--workers 1` and `--output-prefix <data_home>/dataset`, (2) `num_workers=4` with `custom_text_document` -> `--workers 4` and `--output-prefix <data_home>/custom`, and (3) a `data_prefix` without the suffix used as-is.

File: tests/benchmarks/model_benchmarks/test_megatron_gpt.py [hunk 3]
```diff
@@ -174,6 +174,72 @@ def test_megatron_gpt_dataset(self):
...
+        benchmark._preprocess()
+        ret = benchmark._generate_dataset()
+        # Dataset generation will fail because the mocked run_command does not actually
+        # produce .bin/.idx files; we only care about the constructed command.
+        assert ret is False
+        assert mock_run_command.call_count >= 1
+        cmd = mock_run_command.call_args_list[0].args[0]
+        assert '--workers 1 ' in cmd, cmd
+        assert f'--output-prefix {os.path.join(self._tmp_dir, "dataset")} ' in cmd, cmd
+
```
[the remark this answers is not in the document]
Comment (polarG) [reply tests/benchmarks/model_benchmarks/test_megatron_gpt.py]: Addressed in b90dbf964c83bda263c3f7007a0bcc447a59a3b6.

File: tests/benchmarks/model_benchmarks/test_megatron_gpt.py [hunk 4]
```diff
@@ -174,6 +174,72 @@ def test_megatron_gpt_dataset(self):
...
+        # (default data_prefix='dataset_text_document' with the suffix stripped).
+        benchmark = benchmark_cls(
+            self.benchmark_name,
+            parameters=(
+                f'--code_base /root/Megatron-DeepSpeed --data_home {self._tmp_dir} '
+                f'--batch_size 2048 --num_workers 0 '
+                f'--dataset_url http://example.com/data.json'
+            ),
+        )
+        benchmark._preprocess()
```
[the remark this answers is not in the document]
Comment (polarG) [reply tests/benchmarks/model_benchmarks/test_megatron_gpt.py]: Addressed in b90dbf964c83bda263c3f7007a0bcc447a59a3b6.

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 5]
```diff
@@ -651,13 +651,40 @@ def _generate_dataset(self):
...
+                    # data_prefix (stripping the '_text_document' suffix when present) so that
+                    # the generated files match the existence checks above for any custom
+                    # data_prefix value.
+                    output_prefix_basename = self._args.data_prefix
+                    if output_prefix_basename.endswith('_text_document'):
+                        stripped = output_prefix_basename[:-len('_text_document')]
+                        # Guard against data_prefix == '_text_document' which would
+                        # leave an empty basename and produce a malformed --output-prefix
+                        # ending in '/'. Fall back to the original value in that case.
+                        output_prefix_basename = stripped or output_prefix_basename
```
[the remark this answers is not in the document]
Comment (polarG) [reply superbench/benchmarks/model_benchmarks/megatron_gpt3.py]: Addressed in commit 9018df04: added a fail-fast validation in `_generate_dataset()` that requires `data_prefix` to end with `_text_document` and have a non-empty stem when dataset generation is needed (option (a) from your suggestion). Returns `False` with `ReturnCode.DATASET_GENERATION_FAILURE` and a clear error message otherwise.

File: tests/benchmarks/model_benchmarks/test_megatron_gpt.py [hunk 6]
```diff
@@ -174,6 +174,90 @@ def test_megatron_gpt_dataset(self):
...
+        # Case 4: edge case - data_prefix == '_text_document' should NOT strip down
+        # to an empty basename (which would produce '--output-prefix <data_home>/').
+        # Fall back to using '_text_document' as the basename.
+        _run_case(
+            extra_params='--num_workers 1 --data_prefix _text_document',
+            expected_workers=1,
+            expected_prefix_basename='_text_document',
+            expected_data_prefix='_text_document',
+        )
+
```
[the remark this answers is not in the document]
Comment (polarG) [reply tests/benchmarks/model_benchmarks/test_megatron_gpt.py]: Addressed in commit 9018df04: replaced Cases 3 and 4 with negative cases (`_run_invalid_case`) that assert `_preprocess()` returns `False` and `run_command` is never invoked for invalid `data_prefix` values (`mydata` without suffix, `_text_document` with empty stem after stripping). This now matches the contract enforced in `_generate_dataset()`.

**Review by guoshzhao — APPROVED** [review]

**Review by WenqingLan1 — APPROVED** [review]

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 7]
```diff
@@ -651,13 +651,42 @@ def _generate_dataset(self):
...
+                    # when producing the .bin/.idx files. For the existence check below
+                    # (which looks for {data_prefix}.bin/.idx) to pass, data_prefix must end
+                    # with '_text_document' and have a non-empty stem when generation is needed.
+                    suffix = '_text_document'
+                    if not self._args.data_prefix.endswith(suffix) or self._args.data_prefix == suffix:
+                        logger.error(
+                            'data_prefix must end with "{}" and have a non-empty stem when '
+                            'dataset generation is required (got "{}"). preprocess_data.py '
+                            'always appends "{}" to --output-prefix.'.format(suffix, self._args.data_prefix, suffix)
+                        )
```
[the remark this answers is not in the document]
Comment (polarG) [reply superbench/benchmarks/model_benchmarks/megatron_gpt3.py:677]: Addressed in a88294d8: the `data_prefix` suffix validation and the new `num_workers >= 0` check now run **before** `download_file(self._args.dataset_url, ...)` (megatron_gpt3.py L654 / L666 → L674). Invalid prefixes and negative `num_workers` now fail fast without downloading the dataset. The new test case verifies this: `download_file.call_count == 2` (vocab + merges only, no dataset) for the negative `num_workers` rejection path.

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 8]
```diff
@@ -651,13 +651,42 @@ def _generate_dataset(self):
...
+                    # but preprocess_data.py requires workers>=1 for multiprocessing.Pool.
+                    preprocess_workers = max(1, self._args.num_workers)
+                    if preprocess_workers != self._args.num_workers:
+                        logger.warning(
+                            'preprocess_data.py requires --workers >= 1; '
+                            'overriding num_workers={} to {} for dataset preprocessing only '
+                            '(DataLoader still uses num_workers={}).'.format(
+                                self._args.num_workers, preprocess_workers, self._args.num_workers
+                            )
+                        )
```
[the remark this answers is not in the document]
Comment (polarG) [reply superbench/benchmarks/model_benchmarks/megatron_gpt3.py:698]: Addressed in a88294d8: `max(1, num_workers)` was replaced with an explicit validation + a zero-only clamp. Negative `num_workers` is now rejected with `ReturnCode.INVALID_ARGUMENT` and `return False` (megatron_gpt3.py L666–L670). Only `num_workers == 0` is clamped to `preprocess_workers = 1` (because `multiprocessing.Pool(0)` crashes); the DataLoader still receives the user-provided `0`, which is valid. The Megatron training command therefore never sees a negative `--num-workers`. New test case `--num_workers -1` asserts `_preprocess() is False`, `run_command.call_count == 0`, and `return_code == INVALID_ARGUMENT`.

File: tests/benchmarks/model_benchmarks/test_megatron_gpt.py [hunk 9]
```diff
@@ -174,6 +174,100 @@ def test_megatron_gpt_dataset(self):
...
+            def _side_effect(*_args, **_kwargs):
+                for ext in ('.bin', '.idx'):
+                    p = Path(self._tmp_dir) / f'{prefix}{ext}'
+                    p.touch()
+                    created_files.append(p)
+
+            return _side_effect
+
+        self.addCleanup(lambda: [p.unlink() for p in created_files if p.is_file()])
+
```
[the remark this answers is not in the document]
Comment (polarG) [reply tests/benchmarks/model_benchmarks/test_megatron_gpt.py]: Addressed in a88294d8: replaced the side-effect list comprehension with a named helper and `self.addCleanup`:\n\n```python\ndef _cleanup_created_files():\n    for p in created_paths:\n        if p.is_file():\n            p.unlink()\nself.addCleanup(_cleanup_created_files)\n```\n\nNo throwaway list is allocated and the intent reads as cleanup rather than as a comprehension.

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 10]
```diff
@@ -657,15 +657,53 @@ def _generate_dataset(self):
...
+                        self._result.set_return_code(ReturnCode.DATASET_GENERATION_FAILURE)
+                        return False
+
+                    if self._args.num_workers < 0:
+                        logger.error(
+                            'num_workers must be >= 0 (got {}).'.format(self._args.num_workers)
+                        )
+                        self._result.set_return_code(ReturnCode.INVALID_ARGUMENT)
+                        return False
```
[the remark this answers is not in the document]
Comment (polarG) [reply superbench/benchmarks/model_benchmarks/megatron_gpt3.py]: Addressed in f6dad81a. Hoisted the `num_workers < 0` validation to the top of `_generate_dataset()` so it now runs unconditionally — including when the `.bin`/`.idx` files already exist — and uses `ReturnCode.INVALID_ARGUMENT`. A negative `--num_workers` can no longer slip through into the Megatron training command as `--num-workers -1`.

File: superbench/benchmarks/model_benchmarks/megatron_gpt3.py [hunk 11]
```diff
@@ -657,15 +657,53 @@ def _generate_dataset(self):
...
+                    # with '_text_document' and have a non-empty stem when generation is needed.
+                    suffix = '_text_document'
+                    if not self._args.data_prefix.endswith(suffix) or self._args.data_prefix == suffix:
+                        logger.error(
+                            'data_prefix must end with "{}" and have a non-empty stem when '
+                            'dataset generation is required (got "{}"). preprocess_data.py '
+                            'always appends "{}" to --output-prefix.'.format(suffix, self._args.data_prefix, suffix)
+                        )
+                        self._result.set_return_code(ReturnCode.DATASET_GENERATION_FAILURE)
+                        return False
```
[the remark this answers is not in the document]
Comment (polarG) [reply superbench/benchmarks/model_benchmarks/megatron_gpt3.py]: Addressed in f6dad81a. Changed the invalid `data_prefix` branch to set `ReturnCode.INVALID_ARGUMENT` instead of `DATASET_GENERATION_FAILURE`, matching the convention used by the other CLI-arg validations in this benchmark (`code_base`, `micro_batch_size`, `precision`, `num_workers`).

File: tests/benchmarks/model_benchmarks/test_megatron_gpt.py [hunk 12]
```diff
@@ -174,6 +174,120 @@ def test_megatron_gpt_dataset(self):
...
+        def _run_invalid_case(extra_params):
+            """Assert _preprocess() fails fast (no run_command call) for invalid data_prefix."""
+            mock_run_command.reset_mock()
+            mock_run_command.side_effect = None
+            benchmark = _build_benchmark(extra_params)
+            assert benchmark._preprocess() is False
+            assert mock_run_command.call_count == 0
+            assert benchmark.return_code == ReturnCode.DATASET_GENERATION_FAILURE
+
```
[the remark this answers is not in the document]
Comment (polarG) [reply tests/benchmarks/model_benchmarks/test_megatron_gpt.py]: Addressed in f6dad81a. Updated `_run_invalid_case` to assert `ReturnCode.INVALID_ARGUMENT` to match the new return-code contract for invalid `data_prefix`. Also tightened `_run_invalid_workers_case` to assert `mock_download_file.call_count == 0` since `num_workers` is now validated at the top of `_generate_dataset()` (before vocab/merges downloads), so a negative value fails fast with no side effects.
